channeld/channeld.c

```diff
@@ -2199,40 +2586,1304 @@ static void handle_unexpected_reestablish(struct peer *peer, const u8 *msg)
+
+	/* Set the commit_sig on the commitment tx psbt */
+	if (!psbt_input_set_signature(signed_psbt,
+				      splice_funding_index,
+				      their_pubkey,
+				      &their_sig)) {
+
+		status_failed(STATUS_FAIL_INTERNAL_ERROR,
+			      "Unable to set signature internally "
+			      "funding_index: %d "
+			      "pubkey: %s "
+			      "signature: %s "
+			      "psbt: %s",
+			      splice_funding_index,
+			      type_to_string(tmpctx, struct pubkey, their_pubkey),
+			      type_to_string(tmpctx, struct bitcoin_signature, &their_sig),
+			      type_to_string(tmpctx, struct wally_psbt, signed_psbt));
+	}
+
+	psbt_input_set_witscript(signed_psbt,
+				 splice_funding_index,
+				 wit_script);
+
+	final_tx = bitcoin_tx_with_psbt(tmpctx, signed_psbt);
+
+	wit_stack = bitcoin_witness_2of2(signed_psbt, &splice_sig, &their_sig,
+					 &peer->channel->funding_pubkey[LOCAL],
+					 their_pubkey);
+
+	bitcoin_tx_input_set_witness(final_tx, splice_funding_index, wit_stack);
+
+	/* DTODO: validate our peer's signatures are correct
+	 * see closingd.c receive_offer close_tx / check_tx_sig */
+
+	outmsg = towire_channeld_update_inflight(NULL, signed_psbt);
+	wire_sync_write(MASTER_FD, take(outmsg));
+
+	if (!do_i_sign_first(peer, signed_psbt, TX_INITIATOR)) {
+		status_debug("Splice accepter: we sign first");
+		peer_write(peer->pps, take(sigout_msg));
+	}
+
+	/* Sending of TX_SIGNATURE implies the end of stfu mode */
+	end_stfu_mode(peer);
+
+	outmsg = towire_channeld_update_inflight(NULL, signed_psbt);
+	wire_sync_write(MASTER_FD, take(outmsg));
+
+	if (!psbt_finalize(signed_psbt))
+		status_failed(STATUS_FAIL_INTERNAL_ERROR,
+			      "Splice psbt_finalize failed");
+
+	outmsg = towire_channeld_splice_confirmed_signed(tmpctx, final_tx,
+							 chan_output_index);
+	wire_sync_write(MASTER_FD, take(outmsg));
+	send_channel_update(peer, 0);
+
+	reset_splice(peer);
+}
+
+/* This occurs once our 'stfu' transition was successful. */
+static void handle_splice_stfu_success(struct peer *peer)
+{
+	u8 *msg = towire_splice(tmpctx,
+				&peer->channel_id,
+				&chainparams->genesis_blockhash,
+				peer->splice_opener_funding,
+				peer->splice_feerate_per_kw,
+				peer->current_psbt->tx->locktime,
+				&peer->channel->funding_pubkey[LOCAL]);
+	peer_write(peer->pps, take(msg));
+}
+
+/* User has begun a splice with `splice_init` command. Here we request entry
+ * into STFU mode, when we get it, send `splice` to our peer.
+ * Later the peer will send `splice_ack` and the code that starts the actual
+ * splice happens at that point in `splice_intiator()`. */
+static void handle_splice_init(struct peer *peer, const u8 *inmsg)
+{
+	peer->current_psbt = tal_free(peer->current_psbt);
+
+	if (!fromwire_channeld_splice_init(peer, inmsg, &peer->current_psbt,
+					   &peer->splice_opener_funding,
+					   &peer->splice_feerate_per_kw,
+					   &peer->splice_force_feerate))
+		master_badmsg(WIRE_CHANNELD_SPLICE_INIT, inmsg);
+
+	if (peer->on_stfu_success)
```

Comment:
I think the correct code should actually be checking `stfu_request` instead of `on_stfu_success` which hopefully makes it more clear / become self documenting code.